### PR TITLE
Replace sync CommitMessageManager API with async

### LINKED ISSUE #10926: `CommitMessageManager` blocks UI thread
### Environment

- Git Extensions 4.2.0.16609
- Build 76f14e6c102e91355910b3e28a5b093d6add765e (Dirty)
- Git 2.39.2.windows.1 (recommended: 2.40.0 or later)
- Microsoft Windows NT 10.0.22621.0
- .NET 6.0.16
- DPI 96dpi (no scaling)
- Microsoft.WindowsDesktop.App Versions

```
    Microsoft.WindowsDesktop.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 6.0.15 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 6.0.16 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 7.0.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 8.0.0-preview.2.23128.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

### Issue description

The issue is very prominent working with repos located in WSL. Various operations in the Commit dialog lock the dialog (and the app). I've taken various memdumps, and all of those seem to be blocked at the same line:
image
image

### Steps to reproduce

Try to commit in a WSL2 repo.

### Did this work in previous version of GitExtensions?

IDK

### Diagnostics

_No response_

### PR DESCRIPTION
**Pull request by RussKie**
Resolves #10926

The use of sync API leads to observable delays and locking
the application when working with repos located in WSL.

## Test methodology 

- manual
- unit tests

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: GitUI/CommandsDialogs/FormCommit.cs [hunk 1]
```diff
@@ -477,8 +477,11 @@ protected override void OnFormClosed(FormClosedEventArgs e)
                 // a special meaning, and can be dangerous if used inappropriately.
                 if (CommitKind is (CommitKind.Normal or CommitKind.Amend))
                 {
-                    _commitMessageManager.MergeOrCommitMessage = Message.Text;
-                    _commitMessageManager.AmendState = Amend.Checked;
+                    ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
+                    {
+                        await _commitMessageManager.SetMergeOrCommitMessageAsync(Message.Text);
+                        await _commitMessageManager.SetAmendStateAsync(Amend.Checked);
+                    });
```
Comment (mstv) [inline GitUI/CommandsDialogs/FormCommit.cs:484]:
```suggestion
                    ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
                    {
                        await _commitMessageManager.SetMergeOrCommitMessageAsync(Message.Text);
                        await _commitMessageManager.SetAmendStateAsync(Amend.Checked);
                    }).FileAndForget();
```

File: GitUI/CommandsDialogs/FormCommit.cs [hunk 2]
```diff
@@ -1369,10 +1378,10 @@ void DoCommit()
                     {
                         // Save last commit message in settings. This way it can be used in multiple repositories.
                         AppSettings.LastCommitMessage = Message.Text;
-
-                        _commitMessageManager.WriteCommitMessageToFile(Message.Text, CommitMessageType.Normal,
-                                                                       usingCommitTemplate: !string.IsNullOrEmpty(_commitTemplate),
-                                                                       ensureCommitMessageSecondLineEmpty: AppSettings.EnsureCommitMessageSecondLineEmpty);
+                        ThreadHelper.JoinableTaskFactory.RunAsync(
```
Comment (mstv) [inline GitUI/CommandsDialogs/FormCommit.cs:1381]: The file needs to be written before "git commit" is called.
```suggestion
                        ThreadHelper.JoinableTaskFactory.Run(
```

File: GitUI/CommandsDialogs/FormMergeBranch.cs [hunk 3]
```diff
@@ -104,9 +103,10 @@ private void OkClick(object sender, EventArgs e)
             {
                 // [!] Do not reset the last commit message stored in AppSettings.LastCommitMessage
 
-                _commitMessageManager.WriteCommitMessageToFile(mergeMessage.Text, CommitMessageType.Merge,
-                                                               usingCommitTemplate: false,
-                                                               ensureCommitMessageSecondLineEmpty: false);
+                ThreadHelper.JoinableTaskFactory.RunAsync(
```
Comment (mstv) [inline GitUI/CommandsDialogs/FormMergeBranch.cs:106]: ditto
```suggestion
                ThreadHelper.JoinableTaskFactory.Run(
```

File: GitUI/CommandsDialogs/FormCommit.cs (hunk 1 above)
Comment (RussKie) [reply GitUI/CommandsDialogs/FormCommit.cs:484]: That actually has to be sync call, must have been a bad copy paste. Thanks

File: GitUI/CommandsDialogs/FormCommit.cs [hunk 4]
```diff
@@ -477,8 +477,11 @@ protected override void OnFormClosed(FormClosedEventArgs e)
                 // a special meaning, and can be dangerous if used inappropriately.
                 if (CommitKind is (CommitKind.Normal or CommitKind.Amend))
                 {
-                    _commitMessageManager.MergeOrCommitMessage = Message.Text;
-                    _commitMessageManager.AmendState = Amend.Checked;
+                    ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
```
Comment (RussKie) [inline GitUI/CommandsDialogs/FormCommit.cs:480]:
```suggestion
                    ThreadHelper.JoinableTaskFactory.Run(async () =>
```

File: GitCommands/CommitMessageManager.cs [hunk 5]
```diff
@@ -228,13 +248,14 @@ private void WriteFile(string filePath, string errorTitle, string content, Encod
...
             }
-            catch (Exception ex)
+            catch (Exception ex) when (ex is not OperationCanceledException && ex is not ObjectDisposedException)
             {
+                await _owner.SwitchToMainThreadAsync();
+
                 // No need to cancel the other operations in FormCommit - just let the user know that something went wrong
-                MessageBox.Show(null, string.Format(CannotAccessFile, ex.Message, filePath), errorTitle,
-                    MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
+                MessageBox.Show(null, string.Format(CannotAccessFile, ex.Message, filePath), errorTitle, MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
```
Comment (RussKie) [inline GitCommands/CommitMessageManager.cs:258]:
```suggestion
                MessageBox.Show(_owner, string.Format(CannotAccessFile, ex.Message, filePath), errorTitle, MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
```

File: GitCommands/CommitMessageManager.cs [hunk 6]
```diff
@@ -228,13 +248,14 @@ private void WriteFile(string filePath, string errorTitle, string content, Encod
 
             try
             {
-                _fileSystem.File.WriteAllText(filePath, content, encoding ?? Encoding.Default);
+                await _fileSystem.File.WriteAllTextAsync(filePath, content, encoding ?? Encoding.Default, cancellationToken);
             }
-            catch (Exception ex)
+            catch (Exception ex) when (ex is not OperationCanceledException && ex is not ObjectDisposedException)
```
Comment (RussKie) [inline GitCommands/CommitMessageManager.cs:253]:
```suggestion
            catch (Exception ex) when (ex is not (OperationCanceledException or ObjectDisposedException))
```

**Review by mstv — APPROVED** [review]

**Comment by RussKie** [comment]
Thank you

**Pull request merged**